presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcess.java

```diff
@@ -0,0 +1,273 @@
+import com.facebook.presto.spark.execution.property.NativeExecutionSystemConfig;
+import com.facebook.presto.spark.execution.property.WorkerProperty;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.sql.planner.PlanFragment;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.units.Duration;
+import org.apache.spark.SparkFiles;
+
+import javax.annotation.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.URI;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static com.facebook.airlift.http.client.HttpStatus.OK;
+import static com.facebook.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.facebook.presto.spi.StandardErrorCode.NATIVE_EXECUTION_BINARY_NOT_EXIST;
+import static com.facebook.presto.spi.StandardErrorCode.NATIVE_EXECUTION_PROCESS_LAUNCH_ERROR;
+import static com.facebook.presto.spi.StandardErrorCode.NATIVE_EXECUTION_TASK_ERROR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.util.concurrent.Futures.addCallback;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class NativeExecutionProcess
+        implements AutoCloseable
+{
+    private static final Logger log = Logger.get(NativeExecutionProcess.class);
+    private static final String NATIVE_EXECUTION_TASK_ERROR_MESSAGE = "Encountered too many errors talking to native process. The process may have crashed or be under too much load.";
+    private static final String WORKER_CONFIG_FILE = "/config.properties";
+    private static final String WORKER_NODE_CONFIG_FILE = "/node.properties";
+    private static final String WORKER_CONNECTOR_CONFIG_FILE = "/catalog/";
+
+    private final Session session;
+    private final PrestoSparkHttpServerClient serverClient;
+    private final URI location;
+    private final int port;
+    private final TaskManagerConfig taskManagerConfig;
+    private final NativeExecutionSystemConfig systemConfig;
+    private final NativeExecutionNodeConfig nodeConfig;
+    private final NativeExecutionConnectorConfig connectorConfig;
+    private final ScheduledExecutorService errorRetryScheduledExecutor;
+    private final RequestErrorTracker errorTracker;
+    private final HttpClient httpClient;
+    private final NativeExecutionTask nativeExecutionTask;
```

Comment:
I think since logically the NativeExecutionProcess need to  outlast the NativeExecutionTask and IMO it's reasonable to let it controls the lifecycle of the NativeExecutionTask, WDYT?

Reply:
Since we are also referencing NativeExecutionTask in NativeExecutionOperator for calls and controls then operator would be a better place for both task and process. I believe putting task inside process does not actually do lifecycle control as JVM has its own garbage collection mechanism. Even if last reference to process is gone it will just be the process that is garbage collected but not the task within as long as there is something referencing that task. 

Also if we think about higher level concepts, a class being a member is a belong-to relation. Here a task exists more like in parallel with process in a way that process is an abstract concept of cpp process, and it communicates with task through http. Together they get a job done. 
If we compare it to the universal "car" example:
- native process is the car engine.
- native task is the car tire.
- native operator is the car.
having task inside process is like having tire inside engine. But car also references tire as it needs it to roll and run. Now it creates this coupling that is generally discouraged if it could be avoided.

I'm approving this change to unblock since it's been a while. It will be the best to make the change in this PR but I'm okay if we want to do it in a follow-up.

Reply:
IMO, practically NativeExecutionTask strictly depends on the lifecycle of NativeExecutionProcess, if the process has died or not started yet, the task shouldn't be called by the user, giving both the  NativeExecutionTask and NativeExecutionProcess to the user (e.g NativeExecutionOperator) to make sure every time they want to call the task , the native process is alive/available is not ideal - Ideally, we should expose the NativeExecutionTask's APIs through the NativeExecutionProcess and check the process's status before actually calling into the NativeExecutionTask.  I can do a it in the follow-up diff.